FCS witnesses issue

Typechecking the following in FCS with `--langversion:preview` returns an error:
```fs
let test x =
    match x with
    | [| |] -> x
    | _ -> x
```
The error is thrown in [GetWitnessArgs](https://github.com/dotnet/fsharp/blob/main/src/fsharp/symbols/Exprs.fs#L499), resulting in the following exception:
```
[ERROR] The exception has been reported. This internal exception should now be caught at an error recovery point on the stack. Original message: Exception of type 'FSharp.Compiler.ConstraintSolver+ConstraintSolverMissingConstraint' was thrown.)
   at FSharp.Compiler.ErrorLogger.ErrorLoggerExtensions.ErrorLogger.Error[b](ErrorLogger x, Exception exn) in /fsharp/src/fsharp/ErrorLogger.fs:line 414
   at FSharp.Compiler.ErrorLogger.CommitOperationResult[a](OperationResult`1 res) in /fsharp/src/fsharp/ErrorLogger.fs:line 552
   at FSharp.Compiler.SourceCodeServices.FSharpExprConvert.GetWitnessArgs(SymbolEnv cenv, ExprTranslationEnv env, ValRef vref, range m, FSharpList`1 tps, FSharpList`1 tyargs) in /fsharp/src/fsharp/symbols/Exprs.fs:line 499
```

--- a/src/fsharp/QuotationTranslator.fs
+++ b/src/fsharp/QuotationTranslator.fs
@@ -1087,6 +1087,8 @@ and ConvDecisionTree cenv env tgs typR x =
                       | _ ->
                           let ty = tyOfExpr cenv.g e1
                           let eq = mkCallEqualsOperator cenv.g m ty e1 (Expr.Const (Const.Zero, m, ty))
+                          // no need to generate witnesses for generated equality operation calls, see https://github.com/dotnet/fsharp/issues/10389
+                          let env = { env with suppressWitnesses = true }
                           let eqR = ConvExpr cenv env eq
                           QP.mkCond (eqR, ConvDecisionTree cenv env tgs typR dtree, acc)
 

--- a/src/fsharp/QuotationTranslator.fs
+++ b/src/fsharp/QuotationTranslator.fs
@@ -1087,7 +1087,7 @@ and ConvDecisionTree cenv env tgs typR x =
                       | _ ->
                           let ty = tyOfExpr cenv.g e1
                           let eq = mkCallEqualsOperator cenv.g m ty e1 (Expr.Const (Const.Zero, m, ty))
-                          // no need to generate witnesses for generated equality operation calls, see https://github.com/dotnet/fsharp/issues/10389
+                          // no need to generate witnesses for generated equality operation calls, see https://github.com/dotnet/fsharp/issues/10389 
                           let env = { env with suppressWitnesses = true }
                           let eqR = ConvExpr cenv env eq
                           QP.mkCond (eqR, ConvDecisionTree cenv env tgs typR dtree, acc)

--- a/src/fsharp/symbols/Exprs.fs
+++ b/src/fsharp/symbols/Exprs.fs
@@ -725,12 +725,14 @@ module FSharpExprConvert =
                 let elemTy = tyOfExpr cenv.g arg
                 let nullVal = mkNull m elemTy
                 let op = mkCallNotEqualsOperator cenv.g m elemTy arg nullVal
+                let env = { env with suppressWitnesses=true }
                 ConvExprPrim cenv env op
 
             | TOp.ILAsm ([ I_ldlen; AI_conv DT_I4 ], _), _, [arr] -> 
                 let arrayTy = tyOfExpr cenv.g arr
                 let elemTy = destArrayTy cenv.g arrayTy
                 let op = mkCallArrayLength cenv.g m elemTy arr
+                let env = { env with suppressWitnesses=true }
                 ConvExprPrim cenv env op
 
             | TOp.ILAsm ([ I_newarr (ILArrayShape [(Some 0, None)], _)], _), [elemTy], xa ->
@@ -777,8 +779,8 @@ module FSharpExprConvert =
             | TOp.ILAsm ([ ILConvertOp convertOp ], [TType_app (tcref,_)]), _, [arg] -> 
                 let ty = tyOfExpr cenv.g arg
                 let op =
-                    if tyconRefEq cenv.g tcref cenv.g.char_tcr
-                    then mkCallToCharOperator cenv.g m ty arg
+                    if tyconRefEq cenv.g tcref cenv.g.char_tcr then
+                        mkCallToCharOperator cenv.g m ty arg
                     else convertOp cenv.g m ty arg
                 ConvExprPrim cenv env op
 
@@ -1275,8 +1277,10 @@ module FSharpExprConvert =
                             E.IfThenElse (E.TypeTest (tyR, eR) |> Mk cenv m cenv.g.bool_ty, acc, ConvDecisionTree cenv env dtreeRetTy dtree m) 
                         | _ -> 
                             let ty = tyOfExpr cenv.g e1
-                            let eq = mkCallEqualsOperator cenv.g m ty e1 (Expr.Const (Const.Zero, m, ty))
-                            let eqR = ConvExpr cenv env eq 
+                            let eqR =
+                                let eq = mkCallEqualsOperator cenv.g m ty e1 (Expr.Const (Const.Zero, m, ty))
+                                let env = { env with suppressWitnesses = true }
+                                ConvExpr cenv env eq 
                             E.IfThenElse (eqR, ConvDecisionTree cenv env dtreeRetTy dtree m, acc) 
                     | DecisionTreeTest.IsInst (_srcty, tgty) -> 
                         let e1R = ConvExpr cenv env e1

--- a/tests/fsharp/core/quotes/test.fsx
+++ b/tests/fsharp/core/quotes/test.fsx
@@ -4106,7 +4106,6 @@ module QuotationOfComputationExpressionZipOperation =
                                                            [x, y])]))))])),
              PropertyGet (None, builder, []))"""
         
-        
 module CheckEliminatedConstructs = 
     let isNullQuoted (ts : 't[]) =
         <@

--- a/tests/fsharp/core/quotes/test.fsx
+++ b/tests/fsharp/core/quotes/test.fsx
@@ -4107,6 +4107,19 @@ module QuotationOfComputationExpressionZipOperation =
              PropertyGet (None, builder, []))"""
         
         
+module CheckEliminatedConstructs = 
+    let isNullQuoted (ts : 't[]) =
+        <@
+            match ts with
+            | null -> true
+            | _ -> false
+        @>
+
+    let actual1 = ((isNullQuoted [| |]).ToString())
+    checkStrings "brewbreebrvwe1" actual1
+       """IfThenElse (Call (None, op_Equality, [ValueWithName ([||], ts), Value (<null>)]),
+            Value (true), Value (false))"""
+        
 module TestAssemblyAttributes = 
     let attributes = System.Reflection.Assembly.GetExecutingAssembly().GetCustomAttributes(false)
 

--- a/tests/service/ExprTests.fs
+++ b/tests/service/ExprTests.fs
@@ -3509,3 +3509,59 @@ let ``Test ProjectForWitnesses3 GetWitnessPassingInfo`` () =
             argText2 |> shouldEqual "type  ^T ->  ^T ->  ^T"
     end
 
+//---------------------------------------------------------------------------------------------------------
+// This project is for witness arguments, testing for https://github.com/dotnet/fsharp/issues/10364
+
+module internal ProjectForWitnesses4 = 
+
+    let fileSource1 = """
+module M
+
+let isEmptyArray x =
+    match x with
+    | [| |] -> x
+    | _ -> x
+
+let isNull (ts : 't[]) =
+    match ts with
+    | null -> true
+    | _ -> false
+
+let isNullQuoted (ts : 't[]) =
+    <@
+        match ts with
+        | null -> true
+        | _ -> false
+    @>
+
+"""
+
+    let createOptions() = createOptionsAux [fileSource1] ["--langversion:preview"]
+    
+[<Test>]
+let ``Test ProjectForWitnesses4 GetWitnessPassingInfo`` () =
+    let cleanup, options = ProjectForWitnesses4.createOptions()
+    use _holder = cleanup
+    let exprChecker = FSharpChecker.Create(keepAssemblyContents=true)
+    let wholeProjectResults = exprChecker.ParseAndCheckProject(options) |> Async.RunSynchronously
+
+    for e in wholeProjectResults.Diagnostics do 
+        printfn "ProjectForWitnesses4 error: <<<%s>>>" e.Message
+
+    Assert.AreEqual(wholeProjectResults.Diagnostics.Length, 0)
+
+    wholeProjectResults.AssemblyContents.ImplementationFiles.Length |> shouldEqual 1
+    let file1 = wholeProjectResults.AssemblyContents.ImplementationFiles.[0]
+
+    let expected = 
+        ["type M";
+         "let isEmptyArray(x) = (if (if Operators.op_Inequality<'a Microsoft.FSharp.Core.[]> (x,dflt) then Operators.op_Equality<Microsoft.FSharp.Core.int> (ArrayModule.Length<'a> (x),0) else False) then x else x) @ (5,10--5,11)";
+         "let isNull(ts) = (if Operators.op_Equality<'t Microsoft.FSharp.Core.[]> (ts,dflt) then True else False) @ (10,10--10,12)";
+         "let isNullQuoted(ts) = quote((if Operators.op_Equality<'t Microsoft.FSharp.Core.[]> (ts,dflt) then True else False)) @ (15,4--19,6)"]
+
+    let actual = 
+      printDeclarations None (List.ofSeq file1.Declarations)
+      |> Seq.toList 
+    printfn "actual:\n\n%A" actual
+    actual
+      |> shouldPairwiseEqual expected

